### PR TITLE
AF-2432 :  Restoring an old asset version doesn't work outside master branch

### PR DESCRIPTION
**Pull request by akumar074**
* Added CurrentBranch class to track current modifying branch in the editor  

Related PR : 
kie-wb-common [#3238](https://github.com/kiegroup/kie-wb-common/pull/3238)

### CODE REVIEW

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by sthundat** [comment]
Jenkins execute full downstream build

**Comment by akumar074** [comment]
Jenkins retest this please

**Comment by sthundat** [comment]
@akumar07 artifacts are not getting generated.Could you please take a look

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
@sthundat @pefernan , As FDB seems to be failing because of unrelated Test failure, I have added the link to JBPM-server-distribution zip in the description for you to test.

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by sthundat** [comment]
@akumar074 when i select the previous version of an asset from any branch(master/new),getting some server error.Please find the attachment  for details
branchVersionIssue

**Comment by akumar074** [comment]
> @akumar074 when i select the previous version of an asset from any branch(master/new),getting some server error.Please find the attachment for details

@sthundat I followed your steps, but I am unable to reproduce this issue.

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by sthundat** [comment]
Branchrestore
Please check the video

**Comment by akumar074** [comment]
> Please check the video
I followed these steps but I'm still unable to reproduce this.

**Comment by ederign** [comment]
@caponetto can you please take a look at this?

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/common/CurrentBranchImpl.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.uberfire.ext.editor.commons.client.menu.common;
+
+public class CurrentBranchImpl implements CurrentBranch {
+
+    private final String DEFAULT_BRANCH = "master";
```
Comment (sthundat) [inline uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/common/CurrentBranchImpl.java:22]: please make this filed static

Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/common/CurrentBranchImpl.java]: Done

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-backend/src/main/java/org/uberfire/ext/editor/commons/backend/version/VersionServiceImpl.java [hunk 2]
```diff
@@ -70,12 +70,14 @@ public Path getPathToPreviousVersion(String uri) {
...
     public Path restore(final Path _path,
-                        final String comment) {
+                        final String comment,
+                        final String currentBranch) {
         try {
             ioService.startBatch(Paths.convert(_path).getFileSystem());
 
             final org.uberfire.java.nio.file.Path path = pathResolver.resolveMainFilePath(convert(_path));
-            final org.uberfire.java.nio.file.Path target = path.getFileSystem().getPath(path.toString());
+            final org.uberfire.java.nio.file.Path target = path.getFileSystem().getPath(currentBranch,
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-backend/src/main/java/org/uberfire/ext/editor/commons/backend/version/VersionServiceImpl.java]: This is a good suggestion. I have changed the use of getPath from FS to VersionUtil. We can get the same target path now.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-backend/src/main/java/org/uberfire/ext/editor/commons/backend/version/VersionServiceImpl.java [hunk 3]
```diff
@@ -70,12 +70,14 @@ public Path getPathToPreviousVersion(String uri) {
 
     @Override
     public Path restore(final Path _path,
-                        final String comment) {
+                        final String comment,
+                        final String currentBranch) {
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-backend/src/main/java/org/uberfire/ext/editor/commons/backend/version/VersionServiceImpl.java]: I agree `branchName` would be more appropriate for other situations.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/BasicFileMenuBuilder.java [hunk 4]
```diff
@@ -77,7 +78,7 @@ BasicFileMenuBuilder addCopy(final PathProvider provider,
 
     BasicFileMenuBuilder addValidate(final Command command);
 
-    BasicFileMenuBuilder addRestoreVersion(final Path path);
+    BasicFileMenuBuilder addRestoreVersion(final Path path, final CurrentBranch currentBranch);
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/BasicFileMenuBuilder.java]: Done.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/BasicFileMenuBuilderImpl.java [hunk 5]
```diff
@@ -343,8 +344,8 @@ public BasicFileMenuBuilder addValidate(final Command validateCommand) {
     }
 
     @Override
-    public BasicFileMenuBuilder addRestoreVersion(final Path path) {
-        this.restoreCommand = restoreVersionCommandProvider.getCommand(path);
+    public BasicFileMenuBuilder addRestoreVersion(final Path path, final CurrentBranch currentBranch) {
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/BasicFileMenuBuilderImpl.java]: Done.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/common/CurrentBranchImpl.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.uberfire.ext.editor.commons.client.menu.common;
+
+public class CurrentBranchImpl implements CurrentBranch {
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/common/CurrentBranchImpl.java]: Thanks @caponetto for this one. I was unaware of this annotation. It is really helpful.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/BaseEditor.java [hunk 7]
```diff
@@ -792,4 +793,8 @@ public VersionRecordManager getVersionRecordManager() {
     public void setMetadataOriginalHash(final Integer metadataOriginalHash) {
         this.metadataOriginalHash = metadataOriginalHash;
     }
+
+    public void setCurrentBranch(final CurrentBranch currentBranch) {
+        versionRecordManager.setCurrentBranch(currentBranch);
+    }
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/BaseEditor.java]: Removed from the VersionRecordManager class.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java [hunk 8]
```diff
@@ -296,6 +302,14 @@ private void onRestore(final @Observes RestoreEvent restore) {
...
     }
 
+    public void setCurrentBranch(final CurrentBranch currentBranch) {
+        this.currentBranch = currentBranch;
+    }
+
+    public CurrentBranch getCurrentBranch() {
+        return currentBranch;
+    }
+
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java]: Removed from the VersionRecordManager class.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java [hunk 9]
```diff
@@ -226,10 +230,12 @@ private VersionRecord getCurrentVersionRecord() {
     public void restoreToCurrentVersion(boolean withComments) {
         if (withComments)
             restorePopUpPresenter.show(getCurrentPath(),
-                                       getCurrentVersionRecordUri());
+                                       getCurrentVersionRecordUri(),
+                                       getCurrentBranch());
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java]: Done.

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java [hunk 10]
```diff
@@ -226,10 +230,12 @@ private VersionRecord getCurrentVersionRecord() {
...
         else {
             restorePopUpPresenter.restoreCommand(getCurrentPath(),
-                                                 getCurrentVersionRecordUri())
+                                                 getCurrentVersionRecordUri(),
+                                                 getCurrentBranch())
```
[the remark this answers is not in the document]
Comment (akumar074) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/history/VersionRecordManager.java]: Done.

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by akumar074** [comment]
> Hi @akumar074
> I've added some comments on the PRs.
> Please check if they make sense and let me know if you have any questions.
> 
> I see that most of the changes on the PRs is because there is no way of getting the branch name from the `_path` variable in `VersionService#restore` method. We only have the version, and this version could be part of one or more branches. I'm gonna think a little bit more about it but probably we can't avoid these changes.

@caponetto , I have incorporated changes according to your CR. Please find my comments on the same.

**Comment by akumar074** [comment]
Jenkins execute full downstream build please

**Comment by sthundat** [comment]
Functionality is working fine hence approving this PR

**Pull request merged**